archive/imo/imo2021_q1.lean

```diff
@@ -0,0 +1,252 @@
+      apply le_trans _ t,
+      exact le_ceil (sqrt (1 + ↑n) - 1) },
+      suffices : sqrt (4 + 2 * n) ≤ 2 * (sqrt (1 + n) - 3),
+      { linarith },
+    exact radical_inequality hn },
+  { rw hx,
+    exact floor_le _ },
+  { rw [floor_nonneg, le_sub, sub_zero, le_sqrt],
+    all_goals { norm_cast, linarith }},
+
+end
+
+lemma exists_triplet_summing_to_squares (n : ℕ) (hn : n ≥ 100): (∃ (a b c : ℕ),
+  n ≤ a ∧ a < b ∧ b < c ∧ c ≤ 2 * n ∧  (∃ (k : ℕ), a + b = k * k) ∧
+  (∃ (l : ℕ), c + a = l * l) ∧ (∃ (m : ℕ), b + c = m * m)) :=
+begin
+  --If n ≥ 107, we do not explicitly construct the triplet but use an existence
+  --argument from lemma above.
+  by_cases p : n ≥ 107,
+  { have h := exists_numbers_in_interval n p,
+    cases h with l hl,
+    use [2 * l * l - 4 * l, 2 * l * l + 1, 2 * l * l + 4 * l],
+    by_cases p : l > 1,
+    refine ⟨_,⟨_,⟨_,⟨_,⟨_, ⟨_, _⟩⟩⟩⟩⟩⟩,
+    { exact nat.le_sub_right_of_add_le hl.1 },
+    { exact (2 * l * l).sub_lt_succ (4 * l) },
+    { linarith },
+    { exact hl.2 },
+    { use (2 * l - 1),
+      have h₁ : 1 ≤ 2 * l, { linarith },
+      have h₂ : 4 * l ≤ 2 * l * l, { linarith },
+      zify [h₁, h₂],
+      ring_exp_eq },
+    { use (2 * l),
+      have h₁ : 4 * l ≤ 2 * l * l, { linarith },
+      zify [h₁],
+      ring_exp_eq },
+    { use(2 * l + 1),
+      ring_exp_eq },
+      --Otherwise, if 100 ≤ n < 107, then it suffices to consider explicit
+      --construction of a triplet {a, b, c}, which is constructed by setting l=9
+      --in the argument at the start of the file.
+    { exfalso,
+      simp only [not_lt] at p,
+      interval_cases l,
+      all_goals {linarith }}},
+  { refine ⟨126, 163, 198, _, _, _, _, ⟨⟨17, _⟩,⟨⟨18, _⟩,⟨19, _⟩⟩⟩⟩,
+    all_goals { linarith }},
+end
+
+lemma mem_compl_iff_nmem_subset {A : finset ℕ} {a b c n : ℕ}
+ (t : A ⊆ finset.Ico n (2 * n + 2)) (h₁ : n ≤ a)
+ (h₂ : a < b) (h₃ : b < c) (h₄ : c ≤ 2 * n) :
+ (a ∉ A ↔ a ∈ finset.Ico n (2 * n + 2) \ A) ∧
+ (b ∉ A ↔ b ∈ finset.Ico n (2 * n + 2) \ A) ∧
+ (c ∉ A ↔ c ∈ finset.Ico n (2 * n + 2) \ A) :=
+begin
+  refine ⟨⟨_,_⟩, ⟨⟨_,_⟩,⟨_,_⟩⟩⟩,
+  { intro ha,
+    simp only [finset.Ico.mem, finset.mem_sdiff],
+    refine ⟨⟨_,_⟩,ha⟩,
+    all_goals { linarith }},
+  { intro ha,
+    simp only [finset.mem_sdiff, finset.mem_range] at ha,
+    exact ha.2 },
+  { intro hb,
+    simp only [finset.Ico.mem, finset.mem_sdiff],
+    refine ⟨⟨_,_⟩,hb⟩,
+    all_goals { linarith }},
+  { intro hb,
+    simp only [finset.mem_sdiff, finset.mem_range] at hb,
+    exact hb.2 },
+  { intro hc,
+    simp only [finset.Ico.mem, finset.mem_sdiff],
+    refine ⟨⟨_,_⟩,hc⟩,
+    all_goals { linarith }},
+  { intro hc,
+    simp only [finset.mem_sdiff, finset.mem_range] at hc,
+    exact hc.2 },
+end
+
+
+theorem IMO_2021_Q1 : ∀ (n : ℕ), n ≥ 100 → ∀ (A ⊆ finset.Ico n (2 * n + 2)),
+  (∃ (a b : A), a ≠ b ∧ ∃ (k : ℕ), (a:ℕ) + b = k * k) ∨
+  (∃ (a b : finset.Ico n (2 * n + 2) \ A), a ≠ b ∧ ∃(k : ℕ), (a:ℕ) + b = k * k) :=
```

Comment:
Can you factor out a "three things between two piles means two things in the same pile" lemma, that could go in mathlib?

Reply:
I would be happy to do this after I'm finished with this PR :). Although, I'm struggling to think of the right generality at which this should be implemented.

Reply:
Okay -- in my view (happy to be overruled) this is an essential step in finishing this PR. Obviously from your proof it is currently painful to run a "three things in two piles means two things in the same pile" argument. Therefore we should make it less painful, and as the main purpose of IMO PRs is to ensure that mathlib provides the appropriate tools to make IMO arguments run smoothly, this PR needs to be adding this to mathlib per se, not just hacking through it mid-proof here.

Reply:
Thanks for the clarification. As I said before, I struggle to understand what generality this lemma should be implemented in. At any rate, since I'm still a beginner, I'll try to ask around on Zulip when I'm able to construct an mwe.

Reply:
I think either of these would be a good target:

```lean
example {α : Type*} [decidable_eq α] (a x y : finset α) (h : a ⊆ x ∪ y) (ha : 3 ≤ a.card) :
 (∃ b c ∈ a, b ∈ x ∧ c ∈ x) ∨ (∃ b c ∈ a, b ∈ y ∧ c ∈ y) := sorry

example {α : Type*} [decidable_eq α] (a x y : finset α) (h : a ⊆ x ∪ y) {n : ℕ}
  (ha : (2 * n) + 1 ≤ a.card) :
  (∃ b, b ⊆ a ∧ b ⊆ x ∧ n ≤ b.card) ∨ (∃ b, b ⊆ a ∧ b ⊆ y ∧ n ≤ b.card) := sorry
```

Reply:
I might even start by working backwards from the most fundamental fact: If I have a `fintype` with cardinality `2*n+1`, and a `set` in that fintype, then either the set or its complement has cardinality at least `n+1`. Hopefully this or something similar is already in mathlib (let us know if you'd like some help finding it, here or on zulip) with "pigeonhole" in a doc-string nearby. Then some succession of lemmas getting closer and closer to the formulation you want to actually deploy in your proof.

Reply:
I have used @ericrbg's second suggestion in the most recent commit. Do tell me if this still needs some changing. Regardless of that, I am amazed at how much nicer the final proof now looks :smiley:.